sys/psa_crypto/psa_persistent_storage/psa_crypto_cbor_encoder.c

```diff
@@ -0,0 +1,625 @@
+        res = nanocbor_get_uint8(&cbor_key_data, &key_data[pos]);
+        if (res < 0) {
+            goto error;
+        }
+
+        pos += 1;
+    }
+
+    return PSA_SUCCESS;
+
+error:
+    key_len = NULL;
+    key_data = NULL;
+    DEBUG("[persistent storage] error while decoding key: %s",
+                                                nanocbor_error_to_string(res));
+    return nanocbor_error_to_psa_status(res);
+
+}
+
+#if PSA_PROTECTED_KEY_COUNT
+/**
+ * @brief   Decode CBOR encoded protected key
+ *
+ * @param   key     Nanocbor value object containing CBOR encoded key data
+ * @param   slot    Key slot to write decoded key to
+ * @return  psa_status_t
+ */
+static psa_status_t psa_decode_protected_key_slot(nanocbor_value_t *key, psa_key_slot_t *slot)
+{
+    int res = 0;
+
+    psa_key_slot_number_t *slot_no = psa_key_slot_get_slot_number(slot);
+    res = nanocbor_get_uint64(key, slot_no);
+    if (res < 0) {
+        goto error;
+    }
+
+#if IS_USED(MODULE_PSA_ASYMMETRIC)
+    if (PSA_KEY_TYPE_IS_KEY_PAIR(slot->attr.type)) {
+        size_t *pubkey_data_len;
+        uint8_t *pubkey_data;
+
+        psa_get_public_key_data_from_key_slot(slot, &pubkey_data, &pubkey_data_len);
+
+        nanocbor_value_t pub_key;
+        res = nanocbor_enter_array(key, &pub_key);
+        if (res < 0) {
+            goto error;
+        }
+        return psa_decode_key_data(&pub_key, pubkey_data, pubkey_data_len);
+    }
+#endif /* MODULE_PSA_ASYMMETRIC */
+
+    return PSA_SUCCESS;
+
+error:
+    DEBUG("[persistent storage] error while decoding protected key slot: %s",
+                                                nanocbor_error_to_string(res));
+    return nanocbor_error_to_psa_status(res);
+}
+#endif /* PSA_PROTECTED_KEY_COUNT */
+
+#if PSA_ASYMMETRIC_KEYPAIR_COUNT
+/**
+ * @brief   Decode CBOR encoded asymmetric key pair
+ *
+ * @param   key     Nanocbor value object containing CBOR encoded key data
+ * @param   slot    Key slot to write decoded key to
+ * @return  psa_status_t
+ */
+static psa_status_t psa_decode_asymmetric_keypair_slot(nanocbor_value_t *key, psa_key_slot_t *slot)
+{
+    int res;
+    psa_status_t status;
+    size_t *privkey_data_len;
+    size_t *pubkey_data_len;
+    uint8_t *privkey_data;
+    uint8_t *pubkey_data;
+
+    psa_get_key_data_from_key_slot(slot, &privkey_data, &privkey_data_len);
+    psa_get_public_key_data_from_key_slot(slot, &pubkey_data, &pubkey_data_len);
+
+    nanocbor_value_t priv_key;
+    res = nanocbor_enter_array(key, &priv_key);
+    if (res < 0) {
+        goto error;
+    }
+    status = psa_decode_key_data(&priv_key, privkey_data,
+                                    privkey_data_len);
+    if (status != PSA_SUCCESS) {
+        return status;
+    }
+
+    nanocbor_value_t pub_key;
+    nanocbor_skip(key); /* Skip to public key */
+    res = nanocbor_enter_array(key, &pub_key);
+    if (res < 0) {
+        goto error;
+    }
+    return psa_decode_key_data(&pub_key, pubkey_data,
```

Comment:
Same argument as before, I would enter the array in the `psa_decode_key_data` function instead.

Reply:
In case of a single key, the key is not stored in an array, only a bytestring. Arrays are only used, when there are potentially two elements (private and public key, or slot number and public key).
So there's not always an array to enter and we can't use that for everything.